### PR TITLE
fix(kubernetesenv): establish a single watcher per builder

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR attempts to prevent Mixer from creating an ever-growing number of watches against the k8s server for secrets in the case of repeated, regular calls to the `Build()` method of the `kubenetesenv` adapter.

I have not repro'd the issue in question, but here's my diagnosis:

`kubernetesenv.go` contains a block of code for closing handlers as follows:

```
func (h *handler) Close() error {
	for clusterID := range h.builder.controllers {
		_ = h.builder.deleteCacheController(clusterID)
	}
	h.builder.Lock()
	h.builder.kubeHandler = nil
	h.builder.Unlock()

	return nil
}
```

When handlers are removed, the `Close` method will clear *all* of the cached controllers from the *builder*. This includes the local kubeconfig, which I *think* is intentional.

However, the `Build` logic uses the presence of the local kubeconfig controller to determine whether or not set up a watch on the multicluster secrets:

```
path, exists := kubeConfigVar.Lookup()
if !exists {
	path = paramsProto.KubeconfigPath
}
...
_, found := b.controllers[path]
...
if !found {
	if err := initMultiClusterSecretController(b, path, env); err != nil {
		return nil, fmt.Errorf("could not create remote controllers: %v", err)
        }
}
```

In some running clusters, some unfortunate souls are seeing this `Close()` -> `Build()` cycle every ~5m. This means that something like ~300 new watchers are being added per day per Mixer in which the reconfigure cycle is happening (*despite no actual changes in config*).

We have a couple of options:
1) special case the local controller and prevent cleaning during Handler `Close()`. I'm a bit concerned about this option as, iirc, we specifically added that in response to another issue related to leaking `kubernetesenv` controllers
2) ensure that the watch is only setup (successfully) once.
3) ?? something else I'm not thinking of at the moment.

For this PR, I've chosen option 2 and thrown together an example approach.

Note: I've also made the `kubernetesenv` adapter use a singleton builder. So all state for the builder, which consists entirely of the watch should be preserved throughout the lifetime of the Mixer instance.

I'm looking for comments and/or better ideas.

Fixes: https://github.com/istio/istio/issues/19481

### CODE REVIEW

File: mixer/adapter/kubernetesenv/kubernetesenv.go [hunk 1]
```diff
@@ -156,8 +157,16 @@ func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler,
 	b.kubeHandler = &kubeHandler
 
 	if !found {
-		if err := initMultiClusterSecretController(b, path, env); err != nil {
-			return nil, fmt.Errorf("could not create remote controllers: %v", err)
+		// only init a single watcher for multicluster secrets for a given Builder.
+		// if init fails, try again on next Build().
+		// This should prevent an ever-growing number of watchers for a Builder.
+		var initErr error
```
Comment (douglas-reid) [inline mixer/adapter/kubernetesenv/kubernetesenv.go]: might need better locking logic around this section, etc..

**Comment by mandarjog** [comment]
I am wondering why cleanup is not complete.
Cleanup should stop the goroutine and terminate the watches.

**Comment by douglas-reid** [comment]
@mandarjog the builder itself is not ever Closed. it persists across `Build` calls.

**Comment by logicalhan** [comment]
> I am wondering why cleanup is not complete.
> Cleanup should stop the goroutine and terminate the watches.

I agree with this sentiment.

> @mandarjog the builder itself is not ever Closed. it persists across `Build` calls.

But `deleteCacheController` should terminate the watch, no?

**Comment by douglas-reid** [comment]
`deleteCacheController` is *NOT* intended to terminate the watch. It is a handler invoked by the secret controller specifically for removing controllers for *other* clusters. The watch should persist across cache controller adds and deletes.

**Comment by andraxylia** [comment]
Is there any workaround or recovery available for this problem?

**Comment by kyessenov** [comment]
We don't have confirmation that this is responsible for the watcher leak. This is just a best guess.

**Comment by bianpengyuan** [comment]
It looks like secret controller is never properly closed? Doesn't mixer creates a new builder for every new config snapshot?

**Comment by douglas-reid** [comment]
@bianpengyuan I was under the impression for some reason that the Builder itself has the same lifetime as Mixer.  There isn't a hook to clean up Builders, after all. However, spot-checking the `prometheus` adapter, it looks like there is careful work to have a singleton Builder.

I'll make the kubernetesenv use a singleton builder too. I think it is too much work to add change the builder itself to support closing.

Thoughts?

**Comment by mandarjog** [comment]
Modeling in prom singleton sgtm.

File: mixer/adapter/kubernetesenv/kubernetesenv.go (hunk 1 above)
Comment (douglas-reid) [reply mixer/adapter/kubernetesenv/kubernetesenv.go]: nope. above there is a `Lock()` + `defer Unlock()`

**Comment by douglas-reid** [comment]
PTAL. Switched to singleton builder.

File: mixer/adapter/kubernetesenv/kubernetesenv.go [hunk 2]
```diff
@@ -95,8 +96,9 @@ var _ ktmpl.HandlerBuilder = &builder{}
 
 // GetInfo returns the Info associated with this adapter implementation.
 func GetInfo() adapter.Info {
+	singletonBuilder := newBuilder(newKubernetesClient)
```
Comment (logicalhan) [inline mixer/adapter/kubernetesenv/kubernetesenv.go]: Doesn't this initialize a newBuilder with each `GetInfo` call?

Comment (mandarjog) [reply mixer/adapter/kubernetesenv/kubernetesenv.go]: GetInfo is only called during registration

**Comment by douglas-reid** [comment]
@mandarjog PTAL. On config changes, moving to a singleton builder would cause issues, as the handler `Close()` would delete the cached/shared cluster controller for the builder too. I've added a prefix for builder-level controllers in the cache to prevent their removal when a handler is closed.

There might be a more elegant solution, as the whole sharing/caching of controllers has grown as a series of hacks, but I'm trying to avoid reintroducing an old bug in this basically dead code. For instance, there never used to be a 1:1 requirement for builder:handler and some of the code still supports that, but other parts would seem broken if it was 1:n, etc.

**Comment by ssuchter** [comment]
Are we planning on producing a patch release that fixes this soon?

**Comment by douglas-reid** [comment]
@ssuchter that is the idea. i'm also still trying to figure out what the underlying cause is. clearly there is a leak, but this code has been broken since 1.1, if not before... so something has recently changed to dramatically expose this issue.

**Pull request merged**